### PR TITLE
Param requirements

### PR DESCRIPTION
**Pull request by Blacksmoke16**

### CODE REVIEW

**Review by drujensen — APPROVED** [review]
💯

**Review by elorest — APPROVED** [review]
That looks good.

File: README.md [hunk 1]
```diff
@@ -49,45 +58,45 @@ result.params #=> { "post_name" => "my_trip_to_kansas" }
...
 /post/products/23/reviews/
-router: catchall route   1.21M (828.65ns) (± 4.67%)  1.52× slower
- radix: catchall route   1.84M (544.43ns) (± 3.36%)       fastest
+router: catchall route   2.08M (480.64ns) (± 3.67%)  928 B/op   1.72× slower
+radix: catchall route   3.58M (279.59ns) (± 3.24%)  448 B/op        fastest
 
 
 /put/products/Winter-Windproof-Trapper-Hat/dp/B01J7DAMCQ
-globs with suffix match 667.91k (   1.5µs) (± 3.89%) fastest
+globs with suffix match   1.12M (896.18ns) (± 2.20%)  1584 B/op  fastest
```
Comment (robacarp) [inline README.md]: I’d love to see an added metric for the filters you’re adding here

File: README.md [hunk 2]
```diff
@@ -49,45 +58,45 @@ result.params #=> { "post_name" => "my_trip_to_kansas" }
...
-router: deep 986.19k (  1.01µs) (± 4.84%)       fastest
- radix: deep 907.61k (   1.1µs) (± 7.06%)  1.09× slower
+router: deep   1.68M (594.77ns) (± 2.02%)  1041 B/op        fastest
+radix: deep   1.53M (654.95ns) (± 5.03%)   592 B/op   1.10× slower
 
 
 /get/books/23/pages
-router: wrong   1.42M (702.46ns) (± 3.84%)       fastest
- radix: wrong   1.01M (991.22ns) (± 1.30%)  1.41× slower
+router: wrong   2.36M (423.88ns) (± 2.74%)  768 B/op        fastest
```
Comment (robacarp) [inline README.md]: It’s interesting to see the speed difference here, which I’m assuming is due to change in crystal versions

**Comment by robacarp** [comment]
I have a couple thoughts, but I like the idea. I apparently can’t submit my comments on mobile... later I guess. Nice work though. 

Running ameba while also introducing changes makes it a bit hard to see he changes here. Would you mind breaking that out into a different pull?

**Comment by Blacksmoke16** [comment]
@robacarp Sure, can wait till you get a proper review in.

Most of the changes from ameba are specs.

Relevant files for this change would be:
* route_set.cr
* segment.cr
* routes_with_variables_spec.cr

Rest are ameba related only.

File: src/amber/router/segment.cr [hunk 3]
```diff
@@ -5,8 +5,9 @@ module Amber::Router
 
     property route_set : RouteSet(T)
     property segment : String
+    property requirement : Regex?
```
Comment (Sija) [inline src/amber/router/segment.cr]:
```suggestion
    property pattern : Regex?
```

Seems more clear about its nature.

File: src/amber/router/route_set.cr [hunk 4]
```diff
@@ -60,16 +61,16 @@ module Amber::Router
     end
 
     # Add a route to the tree.
-    def add(path, payload : T) : Nil
+    def add(path, payload : T, requirements : Requirements = {} of String => Regex) : Nil
```
Comment (Sija) [inline src/amber/router/route_set.cr]: I'd add handling of requirements as `NamedTuple`.

**Review by robacarp — COMMENTED** [review]
Some of the code paths are a bit vague in my memory, so I may be overlooking something obvious. Is it possible to encapsulate everything around the idea of a requirement to be in the VariableSegment class? Then the #match? method would take care of storing the regex, the logic of checking to see if the parameter matches the regex, etc. What do you think?

File: README.md (hunk 2 above)
Comment (Blacksmoke16) [reply README.md]: Either that or I have a beefier PC than what ran it previously.

**Comment by Blacksmoke16** [comment]
@robacarp Good call.  I added the pattern to only the variable segment class, and just handled the logic from within the `match?` method. 

I also added a section to the benchmark for a route with a requirement.  I should probably add one that fails to match now that i think of it... I'll do that when i get home.

@Sija I renamed the internal property to `pattern`, but left the param on the `add` method as `requirements` as it could include multiple patterns.

File: src/amber/router/route_set.cr (hunk 4 above)
Comment (Blacksmoke16) [reply src/amber/router/route_set.cr]: How would you know the keys, since they are mapped to the named params in the route.

Comment (Sija) [reply src/amber/router/route_set.cr]: I'm not sure I understand. You can always transform it into `Hash` with `.to_h`.

**Comment by Sija** [comment]
Instead of `requirements` I'd suggest using `constraints`, which is more clear and commonly used term.

File: src/amber/router/route_set.cr (hunk 4 above)
Comment (Blacksmoke16) [reply src/amber/router/route_set.cr]: Ohh i see what you mean.  Yea i can prob do that.  Was thinking you wanted the type restriction to be `NamedTuple` and not a hash.  My bad.

File: src/amber/router/segments/variable_segment.cr [hunk 5]
```diff
@@ -1,7 +1,11 @@
 module Amber::Router
   class VariableSegment(T) < Segment(T)
+    def initialize(segment, @pattern : Regex? = nil)
+      super segment
+    end
+
     def match?(segment : String) : Bool
-      true
+      (p = @pattern) ? !(segment =~ p).nil? : true
```
Comment (robacarp) [inline src/amber/router/segments/variable_segment.cr]: This line is a bit hard to parse, can we make it a little more explicit?

```suggestion
      if pattern = @pattern
        segment =~ pattern
      else
        true
      end
```

**Comment by Blacksmoke16** [comment]
@robacarp @Sija This should be good to go now.

File: src/amber/router/route_set.cr (hunk 4 above)
Comment (Sija) [reply src/amber/router/route_set.cr]: I belive this is still left.

**Comment by drujensen** [comment]
@robacarp @Sija need your approval

**Comment by Sija** [comment]
Just `NamedTuple` thing I mentioned in a comment before and it's GTG on my side.

**Comment by Blacksmoke16** [comment]
Ok how about now.

**Comment by Sija** [comment]
Why did you remove `ameba` dependency? Whatever the motivation you should open an issue for that since it's our of scope of this PR.

**Comment by Blacksmoke16** [comment]
It's implemented in #13

**Comment by Sija** [comment]
Oh, I see, fair enough!

**Review by Sija — APPROVED** [review]

**Review by robacarp — APPROVED** [review]
Thanks @Blacksmoke16, this looks great.
